Src/DependencyCollector/Shared/ServiceBusDiagnosticsListener.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            RequestTelemetry telemetry = new RequestTelemetry();
+
+            this.SetCommonProperties(name, payload, activity, telemetry);
+
+            PropertyFetcher urlFetcher = this.GetOrCreatePropertyFetcher(name, EndpointPropertyName);
+
+            // Endpoint is URL of particular ServiceBus, e.g. sb://myservicebus.servicebus.windows.net/
+            telemetry.Url = (Uri)urlFetcher.Fetch(payload);
+
+            PropertyFetcher entityFetcher = this.GetOrCreatePropertyFetcher(name, EntityPropertyName);
+            
+            // We want to make Source field extendable at the beginning as we may add
+            // multi ikey support and also build some special UI for requests coming from the queues.
+            telemetry.Source = "roleName:" + (string)entityFetcher.Fetch(payload);
+
+            this.telemetryClient.TrackRequest(telemetry);
+        }
+
+        private void OnException(string name, object payload)
+        {
+            PropertyFetcher exceptionFetcher = this.GetOrCreatePropertyFetcher(name, ExceptionPropertyName);
+            Exception ex = (Exception)exceptionFetcher.Fetch(payload);
+
+            this.telemetryClient.TrackException(ex);
+        }
+
+        private void SetCommonProperties(string eventName, object eventPayload, Activity activity, OperationTelemetry telemetry)
+        {
+            // activity name looks like 'Microsoft.Azure.ServiceBus.<Name>'
+            // as namespace is too verbose, we'll just take the last node from the activity name as telemetry name
+            string[] activityNameSegments = activity.OperationName.Split('.');
+            if (activityNameSegments.Length > 0)
+            {
+                telemetry.Name = activityNameSegments[activityNameSegments.Length - 1];
+            }
+
+            telemetry.Duration = activity.Duration;
+            telemetry.Timestamp = activity.StartTimeUtc;
+            telemetry.Id = activity.Id;
+            telemetry.Context.Operation.Id = activity.RootId;
+            telemetry.Context.Operation.ParentId = activity.ParentId;
+
+            foreach (var item in activity.Baggage)
+            {
+                if (!telemetry.Context.Properties.ContainsKey(item.Key))
+                {
+                    telemetry.Context.Properties[item.Key] = item.Value;
+                }
+            }
+
+            foreach (var item in activity.Tags)
+            {
+                if (!telemetry.Context.Properties.ContainsKey(item.Key))
+                {
+                    telemetry.Context.Properties[item.Key] = item.Value;
+                }
+            }
+
+            PropertyFetcher statusFetcher = this.GetOrCreatePropertyFetcher(eventName, StatusPropertyName);
+            telemetry.Success = (TaskStatus)statusFetcher.Fetch(eventPayload) == TaskStatus.RanToCompletion;
+        }
+
+        private PropertyFetcher GetOrCreatePropertyFetcher(string eventName, string propertyName)
+        {
```

Comment:
Do the dictionary look-ups need to be protected against multi-thread access? (Should it be a `ConcurrentDictionary`?)

Reply:
well, nothing should break even without a concurrent dictionary. in the worst case, a few fetchers will be created. but anyway, is better to use a concurrent dictionary and all sugars it provides